src/FwdState.cc

```diff
@@ -482,7 +482,9 @@ FwdState::fail(ErrorState * errorState)
     if (pconnRace == racePossible) {
         debugs(17, 5, HERE << "pconn race happened");
         pconnRace = raceHappened;
-        destinations->retryPath(serverConn);
+        if (destinationReceipt.returnable())
+            destinations->returnPath(destinationReceipt);
+        destinationReceipt = nullptr;
```

Comment:
```suggestion
        if (destinationReceipt.returnable()) {
            destinations->returnPath(destinationReceipt);
            destinationReceipt = nullptr;
        }
```

Non-returnable destinationReceipt should imply that it is 'nullptr' already. In other words, if destinationReceipt does not have a position, it does not have a Connection either (destinationReceipt is not obtained from non-destinations sources, such as 'pinned' connections).

Reply:
I agree, but would express that idea using different code:

```suggestion
        if (destinationReceipt) {
            assert(destinationReceipt.returnable());
            destinations->returnPath(destinationReceipt);
            destinationReceipt = nullptr;
        }
```

AFAICT, we never create non-nil PeerConnectionPointers with npos now. Thus, we should remove the returnable() method itself to simplify the interface! Please check for all changes that mention returnable().

Furthermore, I would also assert(destinationReceipt) to noteConnection() where we set destinationReceipt to emphasize that we expect all "positive" HappyConnOpener answers to be returnable -- _we_ decide which destinations can be retried, not HappyConnOpener.